src/tools/tidy/src/deps.rs

```diff
@@ -46,30 +48,37 @@ type ExceptionList = &'static [(&'static str, &'static str)];
 /// * Optionally a tuple of:
 ///     * A list of crates for which dependencies need to be explicitly allowed.
 ///     * The list of allowed dependencies.
+/// * An indication of whether it is a submodule or not.
 // FIXME auto detect all cargo workspaces
-pub(crate) const WORKSPACES: &[(&str, ExceptionList, Option<(&[&str], &[&str])>)] = &[
+pub(crate) const WORKSPACES: &[(&str, ExceptionList, Option<(&[&str], &[&str])>, bool)] = &[
     // The root workspace has to be first for check_rustfix to work.
-    (".", EXCEPTIONS, Some((&["rustc-main"], PERMITTED_RUSTC_DEPENDENCIES))),
+    (".", EXCEPTIONS, Some((&["rustc-main"], PERMITTED_RUSTC_DEPENDENCIES)), false),
     // Outside of the alphabetical section because rustfmt formats it using multiple lines.
     (
         "compiler/rustc_codegen_cranelift",
         EXCEPTIONS_CRANELIFT,
         Some((&["rustc_codegen_cranelift"], PERMITTED_CRANELIFT_DEPENDENCIES)),
+        false,
     ),
     // tidy-alphabetical-start
-    ("compiler/rustc_codegen_gcc", EXCEPTIONS_GCC, None),
+    ("compiler/rustc_codegen_gcc", EXCEPTIONS_GCC, None, false),
     //("library/backtrace", &[], None), // FIXME uncomment once rust-lang/backtrace#562 has been synced back to the rust repo
     //("library/portable-simd", &[], None), // FIXME uncomment once rust-lang/portable-simd#363 has been synced back to the rust repo
     //("library/stdarch", EXCEPTIONS_STDARCH, None), // FIXME uncomment once rust-lang/stdarch#1462 has been synced back to the rust repo
-    ("src/bootstrap", EXCEPTIONS_BOOTSTRAP, None),
-    ("src/ci/docker/host-x86_64/test-various/uefi_qemu_test", EXCEPTIONS_UEFI_QEMU_TEST, None),
+    ("src/bootstrap", EXCEPTIONS_BOOTSTRAP, None, false),
+    (
+        "src/ci/docker/host-x86_64/test-various/uefi_qemu_test",
+        EXCEPTIONS_UEFI_QEMU_TEST,
+        None,
+        false,
+    ),
     //("src/etc/test-float-parse", &[], None), // FIXME uncomment once all deps are vendored
-    ("src/tools/cargo", EXCEPTIONS_CARGO, None),
+    ("src/tools/cargo", EXCEPTIONS_CARGO, None, false),
```

Comment:
I'm not quite clear why this is `false`. `src/tools/cargo` is a submodule.

Overall I'm confused by this PR. tidy has depended on submodules existing for some time. From my understanding, things like the cargo submodule implicitly work via [this check](https://github.com/rust-lang/rust/blob/a3167859f2fd8ff2241295469876a2b687280bdc/src/bootstrap/src/lib.rs#L472-L473) which ensures a subset of submodules are available.

Why not just leave #126225 (or some variation of that), and always check rustc-perf? Is there some significant downside of cloning rustc-perf?

Reply:
@lqd It seemed like you didn't like that we would force contributors to download rustc-perf. Do you have any specific concerns? :)

Reply:
> Overall I'm confused by this PR. tidy has depended on submodules existing for some time. From my understanding, things like the cargo submodule implicitly work via [this check](https://github.com/rust-lang/rust/blob/a3167859f2fd8ff2241295469876a2b687280bdc/src/bootstrap/src/lib.rs#L472-L473) which ensures a subset of submodules are available.

Fetching a dist specific tool and continuously checking its license with tidy on development doesn't seem like a good idea overall, at least in my opinion (I am pretty sure some of the active contributors like @RalfJung don't like this kind of approach either).

Reply:
> I'm not quite clear why this is false. src/tools/cargo is a submodule.

Added automatic submodule detection to prevent issues like this in the future.

cc @Kobzol

Reply:
I think ideally we'd require as few submodules as possible to be checked out. If we can get that number down to 0, that would be perfect. :) But currently that's probably not yet possible.

Reply:
That's fine if there is a significant burden. Would it be possible to update the PR description to explain what the PR changes and the motivation for the change? That way, we're at least clear as to why it is changing (such as "avoids a significant download not relevant to most contributors", or whatever the actual reason is).

Reply:
Updated the PR description.